module/documents/actor/actor.mjs

```diff
@@ -2137,32 +2157,56 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
     if ( Hooks.call("dnd5e.preLongRest", this, config) === false ) return;
 
     if ( config.dialog ) {
-      try { config.newDay = await LongRestDialog.longRestDialog({actor: this}); }
-      catch(err) { return; }
+      try {
+        foundry.utils.mergeObject(config, await LongRestDialog.longRestDialog({actor: this}));
+      } catch(err) { return; }
     }
 
-    return this._rest(config.chat, config.newDay, true);
+    /**
+     * A hook event that fires after a long rest has started, after the configuration is complete.
+     * @function dnd5e.longRest
+     * @memberof hookEvents
+     * @param {Actor5e} actor             The actor that is being rested.
+     * @param {RestConfiguration} config  Configuration options for the rest.
+     * @returns {boolean}                 Explicitly return `false` to prevent the rest from being continued.
+     */
+    if ( Hooks.call("dnd5e.longRest", this, config) === false ) return;
+
+    return this._rest(config);
   }
 
   /* -------------------------------------------- */
 
   /**
    * Perform all of the changes needed for a short or long rest.
    *
-   * @param {boolean} chat           Summarize the results of the rest workflow as a chat message.
-   * @param {boolean} newDay         Has a new day occurred during this rest?
-   * @param {boolean} longRest       Is this a long rest?
-   * @param {number} [dhd=0]         Number of hit dice spent during so far during the rest.
-   * @param {number} [dhp=0]         Number of hit points recovered so far during the rest.
+   * @param {RestConfiguration} config  Configuration data for the rest occurring.
+   * @param {RestResult} [result={}]    Results of the rest operation being built.
+   * @param {*[]} [args]
    * @returns {Promise<RestResult>}  Consolidated results of the rest workflow.
    * @private
    */
-  async _rest(chat, newDay, longRest, dhd=0, dhp=0) {
+  async _rest(config, result={}, ...args) {
+    if ( args.length ) {
+      foundry.utils.logCompatibilityWarning(
+        "Actor5e._rest now takes a config object and a results object as parameters.",
+        { since: "DnD5e 3.1", until: "DnD5e 3.3" }
+      );
+      const [longRest, dhd, dhp] = args;
+      config = { chat: config, newDay: result };
+      config.type = longRest ? "long" : "short";
+      result = { dhd, dhp };
+    }
+
+    if ( foundry.utils.getType(this.system.rest) === "function" ) return this.system.rest(config, result);
```

Comment:
Interesting idea, but wouldn't this mean that each data model that implements this has to be separately responsible for advancing the game time and firing the hooks?

Reply:
Only if they want some logic vastly different than the default actor resting logic, I figure most wouldn't bother and just call the normal rest code. Its not really any different than adding an conditional statement to `Actor#_rest` with special logic for groups, but keeps from filling that method with a bunch of type-specific logic.

Reply:
Well the group actor is kind of a special case where it calls `longRest`/`shortRest` on each group member. But for a more 'normal' Actor, how would that work? If they add a `rest` function to perform some special logic during a rest, they can't call the normal rest code, can they? Because the normal rest code calls `system.rest`, creating an infinite loop?

Reply:
How about this, I've made it so if you return `false` from `ActorDataModel#rest` it will stop the rest of the resting process, but otherwise it will continue.

Reply:
That seems like a good solution to me